k2/csrc/fsa_utils.cu

```diff
@@ -168,16 +207,66 @@ static Fsa TransducerFromStream(std::string first_line, std::istringstream &is,
       float score = scale * StringToFloat(splits[4]);
       arcs.emplace_back(src_state, dest_state, symbol, score);
       state_aux_labels.push_back(aux_label);
+      max_state = std::max(max_state, std::max(src_state, dest_state));
+    } else if (num_fields == 2u) {
+      //   0            1
+      // final_state  score
+      // In this case, openfst is true. There could be multiple final states, so
+      // we first have to collect all the final states, and then work out the
+      // super final state.
+      K2_CHECK(openfst) << "Invalid line: " << line
+                        << "\nFinal state with weight detected in K2 format";
+      original_final_states.push_back(StringToInt(splits[0]));
+      original_final_weights.push_back(StringToFloat(splits[1]));
+      max_state = std::max(max_state, original_final_states.back());
     } else if (num_fields == 1u) {
+      //   0
+      // final_state
       (void)StringToInt(splits[0]);
       break;  // finish reading
     } else {
       K2_LOG(FATAL) << "Invalid line: " << line
-                    << "\nIt expects a line with 5 fields";
+                    << "\nIt expects a line with 1, 2 or 5 fields";
     }
   } while (std::getline(is, line));
 
-  K2_CHECK(is) << "Failed to read fsa from string";
+  // Post processing on final states. When openfst is true, we may have multiple
+  // final states with weights associated with them. We will have to add a super
+  // final state, and convert that into the K2 format (final state with no
+  // weight).
+  if (original_final_states.size() > 0) {
+    K2_CHECK_EQ(openfst, true);
+    K2_CHECK_EQ(original_final_states.size(), original_final_weights.size());
+    int32_t super_final_state = max_state + 1;
+    for (std::size_t = 0; i != original_final_states.size(); ++i) {
+      arcs.emplace_back(original_final_states[i],
+                        super_final_state,
+                        -1,             // kFinalSymbol
+                        scale * original_final_weights[i]);
+      // TODO(guoguo) We are not sure yet what to put as the auxiliary label for
+      //              arcs entering the super final state. The only real choices
+      //              are kEpsilon or kFinalSymbol. We are using kEpsilon for
+      //              now.
+      state_aux_labels.push_back(0);    // kEpsilon
+    }
+  }
+
+  // Sort arcs so that source states are in non-decreasing order. We have to do
+  // this simultaneously for both arcs and auxiliary labels. The following
+  // implementation makes a pair of (Arc, AuxLabel) for sorting.
+  // TODO(guoguo) Optimize this when necessary.
```

Comment:
@chenguoguo 
I find a method with NO memory copying.

```cpp

auto sort_lambda = [&arcs](int32_t i, int32_t j) {
   return arcs[i] < arcs[j];
};

std::sort(arcs.begin(), arc.end(), sort_lambda);
std::sort(state_aux_labels.begin(), state_aux_labels.end(), sort_lambda);
```
This may be helpful if you are adding support for OpenFst.